### PR TITLE
[50259] User multi select on share modal

### PR DESCRIPTION
**Pull request by HDinger**
### Todo

- [x] Allow multiple options in the user_autocompleter of the share WP modal
- [x] Add test cases
- [x] Handle the following cases in the share WP modal
  - [x] One existing user is shared
  - [x] Multiple existing users are shared 
  - [x] One new user created and shared
  - [x] Multiple new users are created and shared
  - [x] Mix of existing and new users are shared
  - [x] Mix of existing, already shared with user, and a new user are shared
- [x] Same cases should work in the members page (as they share the same code base)

known issues
- [x] Update modal when initally there is no share, and then multiple are added at once
- [x] Success message on the members page shows the wrong number

https://community.openproject.org/projects/openproject/work_packages/50259/activity

### CODE REVIEW

File: app/controllers/work_packages/shares_controller.rb [hunk 1]
```diff
@@ -40,25 +40,32 @@ def index
...
-
+    @shares = overall_result.map(&:result)
 
-    if current_visible_member_count > 1
-      respond_with_prepend_share
-    else
-      respond_with_replace_modal
+    unless overall_result.nil?
+      if current_visible_member_count > 1
+        respond_with_prepend_share
```
Comment (aaron-contreras) [inline app/controllers/work_packages/shares_controller.rb]: Let's rename this to

```suggestion
        respond_with_prepend_shares
```

File: spec/features/work_packages/share_spec.rb [hunk 2]
```diff
@@ -44,11 +44,13 @@
   shared_let(:non_shared_project_user) { create(:user, firstname: 'Non Shared Project', lastname: 'User') }
   shared_let(:shared_project_user) { create(:user, firstname: 'Shared Project', lastname: 'User') }
   shared_let(:not_shared_yet_with_user) { create(:user, firstname: 'Not shared Yet', lastname: 'User') }
+  shared_let(:another_not_shared_yet_with_user) { create(:user, firstname: 'Not shared Yet', lastname: 'User') }
```
Comment (aaron-contreras) [inline spec/features/work_packages/share_spec.rb]: Half of the flickering specs I found while local testing come from this:

The two not shared with users have the same name so when searching the autocompleter sometimes it picks `not_shared_yet_with_user` and sometimes `another_not_shared_yet_with_user`
```ruby
  shared_let(:not_shared_yet_with_user) { create(:user, firstname: 'Not shared Yet', lastname: 'User') }
  shared_let(:another_not_shared_yet_with_user) { create(:user, firstname: 'Not shared Yet', lastname: 'User') }
```

We could address this by modifying this user's name to
```suggestion
  shared_let(:another_not_shared_yet_with_user) { create(:user, firstname: 'Not shared Yet 2', lastname: 'User') }
```

This would also make it quite simple to understand what their natural position should be alphabetically if we ever want to set an expectation on it

File: spec/features/work_packages/share_spec.rb [hunk 3]
```diff
@@ -224,6 +227,50 @@ def inherited_member_roles(group:)
...
+
+        share_modal.expect_shared_count_of(7)
+
+        # Groups can be added simultaneously as well
+        share_modal.invite_user([not_shared_yet_with_group, empty_group], 'Comment')
+
+        share_modal.expect_shared_count_of(9)
+
+        share_modal.expect_shared_with(not_shared_yet_with_group, 'Comment', position: 1)
+        share_modal.expect_shared_with(empty_group, 'Comment', position: 2)
```
Comment (aaron-contreras) [inline spec/features/work_packages/share_spec.rb]: This is where the other set of flickering specs surge from and I have some implementation suggestions for the feature:

Due to the way turbo stream actions are processed, they're an instruction set parsed from top to bottom. So if in the autocompleter we select (in order) user A, then user B. The shares get created in the controller in that same order: A, then B. When generating the turbo stream responses this is the same order they come in: prepend A, **then** prepend B. Meaning that we will be prepending users in an inverted order to the one we selected.

I feel it would be nice if the prepending reflects the order in which I added the users to the autocompleter from a UX perspective. If we do so, the tests are also just a tad easier to parse through because we can call precisely what you wrote:
```ruby
        # Selected in this order
        share_modal.invite_user([not_shared_yet_with_group, empty_group], 'Comment')

        share_modal.expect_shared_count_of(9)

        # And get prepended in the exact same order
        share_modal.expect_shared_with(not_shared_yet_with_group, 'Comment', position: 1)
        share_modal.expect_shared_with(empty_group, 'Comment', position: 2)
```

This would require a change to the other multiple invite expectations to match this behavior but would be a lot easier to read then in my opinion.

You can accomplish this behavioral change in the controller quite simply by setting:

```ruby
@shares = overall_result.map(&:result).reverse
```

File: spec/features/work_packages/share_spec.rb [hunk 4]
```diff
@@ -326,6 +375,66 @@ def inherited_member_roles(group:)
...
+      share_modal.expect_shared_count_of(6)
+
+      # Add an existing and a non-existing user to the autocompleter
+      share_modal.select_existing_user not_shared_yet_with_user
+      share_modal.select_not_existing_user_option "hello@world.de"
+
+      share_modal.select_invite_role('View')
+      within share_modal.modal_element do
+        click_button 'Share'
+      end
```
Comment (aaron-contreras) [inline spec/features/work_packages/share_spec.rb]: What do you think about:
```ruby
# Example usages:
share_modal.invite_users([not_shared_yet_with_user, "hello@world.de"], 'View')
share_modal.invite_user(not_shared_yet_with_user, 'Edit')
share_modal.invite_user("hello@world.de", 'Edit')

# in spec/support/components/work_packages/share_modal.rb
def invite_user(users, role_name)
  Array(users).each do |user|
    case user
    when String
      select_not_existing_user_option(user)
    when Principal
      select_existing_user(user)
    end
  end

  select_invite_role(role_name)

  within modal_element do
    click_button 'Share'
  end
end

alias_method :invite_users, :invite_user
```

or if we prefer to keep the actions a bit more granular:

```ruby
# Example usages:
share_modal.autocomplete(not_shared_yet_with_user)
share_modal.autocomplete('hello@world.de')
share_modal.autocomplete([not_shared_yet_with_user, other_user, 'hello@world.de'])
### and then
share_modal.select_invite_role('View')
share_modal.click_share

# in spec/support/components/work_packages/share_modal.rb
def autocomplete(user)
  case user
  when String
    select_not_existing_user_option(user)
  when Principal
    select_existing_user(user)
  end
end

def click_share
  within modal_element do
    click_button 'Share'
  end
end
```
This could reduce the amount of methods we maintain and keep the expectations at a higher level resembling the user interactions we're performing.

File: spec/features/work_packages/share_spec.rb
Comment (aaron-contreras) [inline spec/features/work_packages/share_spec.rb:1]: This PR is not at fault for it but I'd say the `share_spec.rb` is starting to get pretty hard to follow just because of the size of it and the amount of users that we rely on throughout different parts of the spec file in different scenarios.

I'd say it makes sense to consider splitting the "multiple user invites" to a separate spec file (similar to how we did for `bulk_sharing_spec.rb`). That way we can have quite a concentrated context that is not muddied by the rest of the `share_spec.rb`

**Review by aaron-contreras — COMMENTED** [review]
It's shaping out quite nicely. Left some inline comments on solutions to the flickering specs you mentioned were troubling you and some other minor remarks

File: spec/features/work_packages/share_spec.rb (hunk 2 above)
Comment (HDinger) [reply spec/features/work_packages/share_spec.rb]: Ah, good catch! 👍

File: spec/features/work_packages/share_spec.rb (hunk 3 above)
Comment (HDinger) [reply spec/features/work_packages/share_spec.rb]: I agree. The sorting order is much more intuitive, if it matches the order in the autocompleter.

File: spec/features/work_packages/share_spec.rb (hunk 4 above)
Comment (HDinger) [reply spec/features/work_packages/share_spec.rb]: I think the first approach is more readable in the final test. 👍

File: app/controllers/members_controller.rb [hunk 5]
```diff
@@ -38,14 +38,29 @@ def index
   end
 
   def create
-    service_call = create_members
+    overall_result = []
```
Comment (klaustopher) [inline app/controllers/members_controller.rb:41]: Here you are initializing with an empty array, [in the other controller with `nil`](https://github.com/opf/openproject/pull/14033/files#diff-1044ba07b8889f8ff65fb7b1338806f8909221c74d285aadcb659334a3ca169dR43). 

When we use the empty array, we don't need the conditional `if overall_result` in line 48 here and can push directly into the array. I think this makes more sense and we should use that approach in both controllers as it needs one less if conditional

**Review by klaustopher — APPROVED** [review]
Looks good to me. Left one remark about the different handling of the list of results. IMHO it would make sense to use `overall_result = []` in both, leave out the `if overall_result` check and just push the result in there and then change the condition from `overall_result.nil?` to `overall_result.blank?` in the turbo controller

File: app/controllers/members_controller.rb (hunk 5 above)
Comment (HDinger) [reply app/controllers/members_controller.rb:41]: Good catch! Thanks 👍
